### ISSUE TITLE
Allow multiple instrumentators for multiple servers

### ISSUE THREAD

**Issue by nikolaikopernik** [trallnag/prometheus-fastapi-instrumentator#29]
# What is the problem?
If I run integration tests with my fastapi server instrumented with metrics I have the following error:
```
ValueError: Duplicated timeseries in CollectorRegistry: {'http_requests', 'http_requests_created', 'http_requests_total'}
```
# How to reproduce? 
```
from fastapi import FastAPI
from prometheus_fastapi_instrumentator import Instrumentator

def test_expose_twice():
    app = FastAPI()
    Instrumentator().instrument(app).expose(app, endpoint="/metrics_1")
    Instrumentator().instrument(app).expose(app, endpoint="/metrics_2")
```

# Issue
The issue is that the Instrumentator always use prometheus_client.REGISTRY for its metrics. When we run our tests we construct multiple instances of our servers so the instrumentator tries to register the same metrics for all of them.

# Possible solution
Accept registry as a constructor parameter to `PrometheusFastApiInstrumentator` class so users can specify their own registries.

**Comment by ponponon** [comment]
# 中文版本

可以使用如下的方式：

```python
from loggers import logger
from fastapi import FastAPI, Form
import uvicorn
from prometheus_fastapi_instrumentator import Instrumentator


app = FastAPI()


@app.get('/')
def root():
    return {"message": "Hello World"}


@app.on_event("startup")
async def startup_event():
    Instrumentator().instrument(app).expose(app)


if __name__ == '__main__':
    uvicorn.run("api:app", host="0.0.0.0", port=5002,)
```
> 把该文件命名为 `api.py`

然后使用 `python api.py` 运行该文件，将获得如下的输出：
```shell
─➤  python api.py                                                       
INFO:     Started server process [14004]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:5002 (Press CTRL+C to quit)
```

具体可以参考:[Instrumentator appears to be broken with the latest FastAPI](https://github.com/trallnag/prometheus-fastapi-instrumentator/issues/80)



--------



# English Version

You can use the following.

```python
from loggers import logger
from fastapi import FastAPI, Form
import uvicorn
from prometheus_fastapi_instrumentator import Instrumentator


app = FastAPI()


@app.get('/')
def root():
    return {"message": "Hello World"}


@app.on_event("startup")
async def startup_event():
    Instrumentator().instrument(app).expose(app)


if __name__ == '__main__':
    uvicorn.run("api:app", host="0.0.0.0", port=5002,)
```


> name the file `api.py`

Then run the file with `python api.py` and you will get the following output.
```shell
─➤  python api.py                                                       
INFO:     Started server process [14004]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:5002 (Press CTRL+C to quit)
```

For details, see:[Instrumentator appears to be broken with the latest FastAPI](https://github.com/trallnag/prometheus-fastapi-instrumentator/issues/80)

**Comment by abhishekvalsan-joveo** [comment]
is there a solution to this?

**Comment by SebastianScherer88** [comment]
I have a somewhat related issue, so adding to this thread - let me know if opening a separate issue would be better. the issue:

I have a fastapi app that i annotate and expose with the instrumentator, and it works as expected in that it adds a `/metrics` `GET` endpoint with all metrics. When i serve this app using more than 1 (lets say uvicorn) servers, i naturally get more than one `/metrics` endpoint. which of the various works i connect to when visiting `{url}/metrics` seems to be randomly decided by the load balancing logic of the parent worker process.

this however is an issue, as prometheus presumably would only ever connect to one randomly selected worker process' `/metrics` endpoint and scrape what is essentialy the `nth` part of the metrics it should get.

is there a way to avoid this (breaking imo) issue and consolidate the metrics resulting from exposing the app to the instrumentator into one metrics endpoint? I have seen other libraries' approach to this (e.g. `kserve` - they dont use this library and rely on the more low level `prometheus-client` library directly), but would be interested in your opinion

**Comment by trallnag** [comment]
@SebastianScherer88, please check out the following section in the documentation of `prometheus-client`:

https://github.com/prometheus/client_python#multiprocess-mode-eg-gunicorn

The gist is that you have to set the environment variable `PROMETHEUS_MULTIPROC_DIR` before starting the respective Python program.

**Comment by navolotsky** [comment]
It seems like the issue fixed in #208 

https://github.com/trallnag/prometheus-fastapi-instrumentator/blob/master/CHANGELOG.md#5100--2023-02-26

**Comment by trallnag** [comment]
Thanks for pointing this out @navolotsky
